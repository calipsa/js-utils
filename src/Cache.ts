import Singleton from './Singleton'

export default class Cache<K, V> extends Singleton {
  map = new Map<K, V>()

  add(idx: K, item: V) {
    this.map.set(idx, item)
  }

  get(idx: K) {
    return this.map.get(idx)
  }

  clear() {
    this.map.clear()
  }
}
