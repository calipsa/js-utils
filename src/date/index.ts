import fromNow from './fromNow'

export {
  // eslint-disable-next-line import/prefer-default-export
  fromNow,
}
