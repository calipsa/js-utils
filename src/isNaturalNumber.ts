/**
 * Checks if a value is a natural number,
 * i.e. is an integer greater than zero
 */
export default (num: any): num is number =>
  Number.isInteger(num) && num > 0
