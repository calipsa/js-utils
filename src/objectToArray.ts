import isNonNegativeInteger from './isNonNegativeInteger'

type O = Readonly<{
  [k: number]: any,
}>

/**
 * Takes all values matching non-negative integer keys in an object
 * and puts them in an array
 */
export default <Obj extends O, V extends Obj[keyof Obj]>(o: Obj) => {
  const indices = Object.keys(o)
    .map(Number)
    .filter(isNonNegativeInteger)

  const minKey = Math.min(...indices)
  const maxKey = Math.max(...indices)
  const arr: V[] = []

  for (let i = minKey; i <= maxKey; ++i) {
    arr[i] = o[i]
  }

  return arr
}
