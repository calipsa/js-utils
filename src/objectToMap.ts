/**
 * Converts an object to a `Map`
 */
// eslint-disable-next-line @typescript-eslint/ban-types
export default <T extends {}, K extends keyof T>(o: T) =>
  new Map<string, T[K]>(Object.entries(o))
