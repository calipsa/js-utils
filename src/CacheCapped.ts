import Cache from './Cache'

export default class CacheCapped<K, V> extends Cache<K, V> {
  private capacity: number

  constructor(capacity: number) {
    super()
    this.capacity = capacity
  }

  rm(item: K) {
    this.map.delete(item)
  }

  add(idx: K, item: V) {
    if (this.map.size === this.capacity && !this.map.has(idx)) {
      const oldest = this.map.keys().next().value
      this.rm(oldest)
    }
    this.map.set(idx, item)
  }
}
