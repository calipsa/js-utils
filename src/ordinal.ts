import isNaturalNumber from './isNaturalNumber'

export default (i: number) => {
  if (!isNaturalNumber(i)) {
    throw new Error(`Number must be a natural number. Instead got: ${i}`)
  }

  const j = i % 10
  const k = (i % 100) - j
  if (k !== 10) {
    if (j === 1) {
      return `${i}st`
    }
    if (j === 2) {
      return `${i}nd`
    }
    if (j === 3) {
      return `${i}rd`
    }
  }
  return `${i}th`
}
