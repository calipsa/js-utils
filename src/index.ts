import * as date from './date'
import * as timezone from './timezone'

import Cache from './Cache'
import CacheCapped from './CacheCapped'
import delay from './delay'
import hasDuplicates from './hasDuplicates'
import isNonNegativeInteger from './isNonNegativeInteger'
import objectToArray from './objectToArray'
import objectToFunction from './objectToFunction'
import objectToFunctionWithBind from './objectToFunctionWithBind'
import queryObjectToString from './queryObjectToString'
import Singleton from './Singleton'
import sliceRotate from './sliceRotate'
import toNearestByStep from './toNearestByStep'

export {
  date,
  timezone,

  Cache,
  CacheCapped,
  delay,
  hasDuplicates,
  isNonNegativeInteger,
  objectToArray,
  objectToFunction,
  objectToFunctionWithBind,
  queryObjectToString,
  Singleton,
  sliceRotate,
  toNearestByStep,
}
