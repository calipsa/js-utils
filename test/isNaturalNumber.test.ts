import { isNaturalNumber } from '../src'

import numbers from './data/numbers'

describe('isNaturalNumber', () => {
  for (const n of numbers) {
    it(`Should isNaturalNumber(${n}) correctly`, () => {
      const result = isNaturalNumber(n)
      expect(result).toMatchSnapshot()
    })
  }
})
