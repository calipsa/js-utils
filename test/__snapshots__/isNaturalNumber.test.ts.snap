// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isNaturalNumber Should isNaturalNumber() correctly 1`] = `false`;

exports[`isNaturalNumber Should isNaturalNumber(-0.4) correctly 1`] = `false`;

exports[`isNaturalNumber Should isNaturalNumber(-1) correctly 1`] = `false`;

exports[`isNaturalNumber Should isNaturalNumber(-4.5) correctly 1`] = `false`;

exports[`isNaturalNumber Should isNaturalNumber(-9007199254740991) correctly 1`] = `false`;

exports[`isNaturalNumber Should isNaturalNumber(-Infinity) correctly 1`] = `false`;

exports[`isNaturalNumber Should isNaturalNumber(0) correctly 1`] = `false`;

exports[`isNaturalNumber Should isNaturalNumber(0) correctly 2`] = `false`;

exports[`isNaturalNumber Should isNaturalNumber(1) correctly 1`] = `true`;

exports[`isNaturalNumber Should isNaturalNumber(1.4142135623730951) correctly 1`] = `false`;

exports[`isNaturalNumber Should isNaturalNumber(1.7976931348623157e+308) correctly 1`] = `true`;

exports[`isNaturalNumber Should isNaturalNumber(2.220446049250313e-16) correctly 1`] = `false`;

exports[`isNaturalNumber Should isNaturalNumber(5) correctly 1`] = `false`;

exports[`isNaturalNumber Should isNaturalNumber(5e-324) correctly 1`] = `false`;

exports[`isNaturalNumber Should isNaturalNumber(10000) correctly 1`] = `true`;

exports[`isNaturalNumber Should isNaturalNumber(9007199254740991) correctly 1`] = `true`;

exports[`isNaturalNumber Should isNaturalNumber(Infinity) correctly 1`] = `false`;

exports[`isNaturalNumber Should isNaturalNumber(NaN) correctly 1`] = `false`;

exports[`isNaturalNumber Should isNaturalNumber(false) correctly 1`] = `false`;

exports[`isNaturalNumber Should isNaturalNumber(null) correctly 1`] = `false`;

exports[`isNaturalNumber Should isNaturalNumber(string) correctly 1`] = `false`;

exports[`isNaturalNumber Should isNaturalNumber(true) correctly 1`] = `false`;

exports[`isNaturalNumber Should isNaturalNumber(undefined) correctly 1`] = `false`;
